src/ccache/execute.cpp

```diff
@@ -299,30 +303,46 @@ execute(Context& ctx,
 {
   LOG("Executing {}", util::format_argv_for_logging(argv));
 
-  {
-    SignalHandlerBlocker signal_handler_blocker;
-    ctx.compiler_pid = fork();
+  int result;
+  posix_spawn_file_actions_t file_actions;
+  if ((result = posix_spawn_file_actions_init(&file_actions))) {
+    throw core::Fatal(
+      FMT("posix_spawn_file_actions_init failed: {}", strerror(result)));
   }
 
-  if (ctx.compiler_pid == -1) {
-    throw core::Fatal(FMT("Failed to fork: {}", strerror(errno)));
+  if ((result = posix_spawn_file_actions_adddup2(
+         &file_actions, *fd_out, STDOUT_FILENO))
+      || (result = posix_spawn_file_actions_addclose(&file_actions, *fd_out))
+      || (result = posix_spawn_file_actions_adddup2(
+            &file_actions, *fd_err, STDERR_FILENO))
+      || (result = posix_spawn_file_actions_addclose(&file_actions, *fd_err))) {
+    throw core::Fatal(FMT("posix_spawn_file_actions_addclose/dup2 failed: {}",
+                          strerror(result)));
   }
 
-  if (ctx.compiler_pid == 0) {
-    // Child.
-    dup2(*fd_out, STDOUT_FILENO);
-    fd_out.close();
-    dup2(*fd_err, STDERR_FILENO);
-    fd_err.close();
-    exit(execv(argv[0], const_cast<char* const*>(argv)));
+  pid_t pid;
+  extern char** environ;
+  result = posix_spawn(&pid,
+                       argv[0],
+                       &file_actions,
+                       nullptr,
+                       const_cast<char* const*>(argv),
+                       environ);
+
```

Comment:
If the process receives a signal here (after `posix_spawn` but before `SignalHandlerBlocker` then the child won't be killed by the signal handler. Unless there's a problem with doing so I suggest moving `posix_spawn` to inside the signal handler blocker block, just like the `fork` call was before.

Reply:
Good point, fixed.